### PR TITLE
Add benchmark for quadratic hash performance

### PR DESCRIPTION
**Pull request by natecook1000**
This is a hopefully reasonably quick benchmark for the issue described in [SR-3268](https://bugs.swift.org/browse/SR-3268). There's a fix in ca99ee2c342e07a277938ba2bb1b5bfd418ce0c2 that can be tested once the benchmark is in.

(This includes an update to the CMakeList template, since the autogenerated file was edited in a previous commit.)

### CODE REVIEW

**Comment by lplarson** [comment]
@natecook1000 Your CMakeLists.txt_template changes aren't in the actual CMakeLists.txt file. Run the scripts/generate_harness/generate_harness.py script to generate it from the template.

**Comment by natecook1000** [comment]
The problem was that someone made changes to the generated version and checked that in — my changes to the template only reflect that change.

**Comment by lplarson** [comment]
@swift-ci benchmark

**Comment by lplarson** [comment]
+ /Users/buildnode/jenkins/workspace/swift-PR-osx-perf/Ninja-Release/swift-macosx-x86_64/bin/Benchmark_Driver compare --log-dir /Users/buildnode/jenkins/workspace/swift-PR-osx-perf/Ninja-Release/swift-macosx-x86_64/benchmark/logs --swift-repo /Users/buildnode/jenkins/workspace/swift-PR-osx-perf/swift --compare-script /Users/buildnode/jenkins/workspace/swift-PR-osx-perf/swift/benchmark/scripts/compare_perf_tests.py
Traceback (most recent call last):
  File "/Users/buildnode/jenkins/workspace/swift-PR-osx-perf/swift/benchmark/scripts/compare_perf_tests.py", line 371, in <module>
    sys.exit(main())
  File "/Users/buildnode/jenkins/workspace/swift-PR-osx-perf/swift/benchmark/scripts/compare_perf_tests.py", line 154, in main
    ratio = (old_results[key] + 0.001) / (new_results[key] + 0.001)
KeyError: 'MapReduceAnyCollectionShort'
Traceback (most recent call last):
  File "/Users/buildnode/jenkins/workspace/swift-PR-osx-perf/swift/benchmark/scripts/compare_perf_tests.py", line 371, in <module>
    sys.exit(main())
  File "/Users/buildnode/jenkins/workspace/swift-PR-osx-perf/swift/benchmark/scripts/compare_perf_tests.py", line 154, in main
    ratio = (old_results[key] + 0.001) / (new_results[key] + 0.001)
KeyError: 'MapReduceAnyCollectionShort'
branch/branch comparison skipped: no previous branch logs
Comparing master/Benchmark_O-20170227132347.log HEAD/Benchmark_O-20170227144206.log ...
Comparing master/Benchmark_Onone-20170227133644.log HEAD/Benchmark_Onone-20170227145919.log ...
-- check-swift-benchmark-macosx-x86_64 finished --

**Comment by natecook1000** [comment]
@lplarson I don't think it knows how to calculate the benchmark comparison when there's a new one added; had the same problem with #6382.

**Comment by natecook1000** [comment]
@swift-ci Please smoke test

**Comment by lplarson** [comment]
You're right. LGTM.

File: benchmark/single-source/HashQuadratic.swift [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        for i in 0..<size {
+            dict1[i] = i * 2
+        }
+
+        var dict2: [Int: Int] = [:]
+        for (k, v) in dict1 {
+            dict2[k] = v
+        }
+    
+        CheckResults(dict2[size/2] == dict2[size/2],
```
Comment (lplarson) [inline benchmark/single-source/HashQuadratic.swift]: @natecook1000 My apologies for missing this. It appears you're trying to compare dict1 and dict2 but you have dict2 on both sides.

Comment (natecook1000) [inline benchmark/single-source/HashQuadratic.swift]: You're so right! I think that will still defeat inlining, but I'll fix ASAP. Thanks!

**Comment by palimondo** [comment]
@natecook1000:
> This is a hopefully reasonably quick benchmark

On my Late 2008 MBP, it takes 40s to run 1 iteration in O build and **640s** in Onone build! For 20 iterations, my machine spends more than 2 hours on this single performance test. Could we adjust the size to something more reasonable? It should take less than a second to run one iteration of test run, the benchmark driver will up the N in the outer loop as necessary to run for about a second.

// cc @lplarson @dabrahams

**Comment by palimondo** [comment]
The CI machine is about 10x faster than my MBP, but it still takes 6s in O, and 50s in Onone build per iteration. That is overall about 20 minutes on this single test.

**Comment by natecook1000** [comment]
The idea was to actually fix this issue in short order so the test wouldn't actually be quadratic. We definitely don't need to be slowing down the entire benchmarking suite if we don't have a fix in flight.

**Comment by palimondo** [comment]
Umm... you've referenced fix in commit https://github.com/apple/swift/commit/ca99ee2c342e07a277938ba2bb1b5bfd418ce0c2, did that not land? I'm confused about where that one commit comes from...

**Comment by palimondo** [comment]
It seems it is in the tree... did that not fix the issue?

**Pull request merged**